Stats: Add option to enable new Stats experience

kangzj opened the pull request:
#### Changes proposed in this Pull Request:
The PR introduces an Stats options named `calypso_stats` to control the experience of Stats, which will easily allow user to preview the new experience if they want  and also would be easier for internal testing etc.

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion
n/a

#### Does this pull request change what data or activity we track or use?
no

#### Testing instructions:
- Open a site with Jetpack plugin and fully connected
- Open `/wp-admin/admin.php?page=jetpack#/settings?term=Jetpack%20Stats`
- Expand the section
- Enable 'Preview new Jetpack Stats experience (Experimental)'
- Open `/wp-admin/admin.php?page=stats`
- Ensure you get the Calypso Stats experience
- Go back and disable the option
- Ensure you get the Old Stats experience

![image](https://user-images.githubusercontent.com/1425433/202062296-fc778ad6-aa6f-4c72-8298-8551ea2ada25.png)

kangzj committed:
added an stats option calypso_stats to control the new Calypso experience

kangzj committed:
changelog

kangzj committed:
fix jetpack changelog to enhancement

jsnmoon commented on projects/plugins/jetpack/_inc/client/traffic/site-stats.jsx:
```diff
@@ -243,6 +243,18 @@ class SiteStatsComponent extends React.Component {
 							</CompactFormToggle>
 						</FormFieldset>
+						<FormFieldset>
+							<CompactFormToggle
+								checked={ !! this.props.getOptionValue( 'calypso_stats' ) }
+								disabled={ ! isStatsActive || unavailableInOfflineMode }
+								toggling={ this.props.isSavingAnyOption( [ 'stats', 'calypso_stats' ] ) }
+								onChange={ this.handleStatsOptionToggle( 'calypso_stats' ) }
+							>
+								<span className="jp-form-toggle-explanation">
+									{ __( 'Preview new Jetpack Stats experience (Experimental)', 'jetpack' ) }
+								</span>
```
```suggestion
							<CompactFormToggle
								checked={ !! this.props.getOptionValue( 'enable_calypso_stats' ) }
								disabled={ ! isStatsActive || unavailableInOfflineMode }
								toggling={ this.props.isSavingAnyOption( [ 'stats', 'enable_calypso_stats' ] ) }
								onChange={ this.handleStatsOptionToggle( 'enable_calypso_stats' ) }
							>
								<span className="jp-form-toggle-explanation">
									{ __( 'Preview the new Jetpack Stats experience (Experimental).', 'jetpack' ) }
								</span>
```

jsnmoon commented on projects/plugins/jetpack/_inc/lib/class.core-rest-api-endpoints.php:
```diff
@@ -2730,6 +2730,13 @@ public static function get_updateable_data_list( $selector = '' ) {
 				'validate_callback' => __CLASS__ . '::validate_boolean',
 				'jp_group'          => 'stats',
 			),
+			'calypso_stats'                        => array(
```
```suggestion
			'enable_calypso_stats'                        => array(
```

jsnmoon commented on projects/plugins/jetpack/_inc/lib/class.core-rest-api-endpoints.php:
```diff
@@ -2730,6 +2730,13 @@ public static function get_updateable_data_list( $selector = '' ) {
 				'jp_group'          => 'stats',
 			),
+			'calypso_stats'                        => array(
+				'description'       => esc_html__( 'Preview new Jetpack Stats experience (Experimental).', 'jetpack' ),
```
```suggestion
				'description'       => esc_html__( 'Preview the new Jetpack Stats experience (Experimental).', 'jetpack' ),
```

jsnmoon commented on projects/plugins/jetpack/_inc/lib/core-api/class.jetpack-core-api-module-endpoints.php:
```diff
@@ -865,6 +865,7 @@ public function update_data( $request ) {
 					break;
 
 				case 'admin_bar':
+				case 'calypso_stats':
```
```suggestion
				case 'enable_calypso_stats':
```

jsnmoon commented on projects/plugins/jetpack/changelog/add-option-to-enable-modern-stats:
```diff
@@ -0,0 +1,4 @@
+Significance: minor
+Type: enhancement
+
+Stats: add stats option `calypso_stats` to allow user get the new experience
```
```suggestion
Stats: add stats option `enable_calypso_stats` to allow users to enable the new Calypso Stats experience
```

jsnmoon commented on projects/plugins/jetpack/modules/stats.php:
```diff
@@ -286,7 +286,7 @@ function stats_admin_menu() {
 
 	// phpcs:ignore WordPress.Security.NonceVerification.Recommended
-	if ( ! isset( $_GET['calypso_stats'] ) ) {
+	if ( ! Stats_Options::get_option( 'calypso_stats' ) ) {
```
```suggestion
	if ( ! Stats_Options::get_option( 'enable_calypso_stats' ) ) {
```

jsnmoon commented on projects/packages/stats-admin/changelog/add-option-to-enable-modern-stats:
```diff
@@ -0,0 +1,4 @@
+Significance: minor
+Type: added
+
+Stats: add stats option `calypso_stats` to allow user get the new experience
```
```suggestion
Stats: add stats option `enable_calypso_stats` to allow users to enable the new Calypso Stats experience
```

jsnmoon reviewed (dismissed):
I added several suggestions to rename `calypso_stats` into `enable_calypso_stats` for improved clarity, but everything looks good otherwise.

Here's an unfortunate (and perhaps hilarious) bug I found for the old Jetpack Stats dashboard: 

1. Disable `enable_calypso_stats` (or `calypso_stats`).
2. Navigate to `/wp-admin/admin.php?page=stats#!/stats/day/`. (In other words, navigate to the old stats page using the new stats URL.)
3. Witness this monstrosity:

<img width="1792" alt="image" src="https://user-images.githubusercontent.com/4044428/202275919-745a1b6b-fa8b-4f24-80c2-838a010f1879.png">

Let's fix this bug in a future PR.

jsnmoon reviewed (commented):
Though on second thought, this change essentially soft-launches our currently broken Calypso/Jetpack Stats (Calypack Stats?) to the public. Are we sure we want to do this?

Perhaps we ought to apply `display: none;` on the new form fieldset? What do you think?

kangzj committed:
rename calypso_stats to enable_calypso_stats

kangzj committed:
changelog

kangzj committed:
tweak option description

kangzj commented:
> I added several suggestions to rename calypso_stats into enable_calypso_stats for improved clarity, but everything looks good otherwise.

Addressed 😄 `enable_calypso_stats` was in my initial commits but I thought I might follow the `admin_bar` style, so changed to `calypso_stats`.

> Though on second thought, this change essentially soft-launches our currently broken Calypso/Jetpack Stats (Calypack Stats?) to the public. Are we sure we want to do this?

Yep good point. Now the option is hidden.

> Witness this monstrosity

That's some dead loop [for not dealing with the hash part](https://github.com/Automattic/jetpack/blob/dc4a5696f2ca183def49a860fc6ddbbffc9483d6/projects/plugins/jetpack/modules/stats.php#L396) 😅  Not related to the PR tho, I'll address separately.
